src/shared/metrics.h

```diff
@@ -0,0 +1,133 @@
+
+#define METRICS_VTABLE_START()                                          \
+        {                                                               \
+                .type = _METRICS_VTABLE_START,                          \
+                .x = {                                                  \
+                        .start = {                                      \
+                                .element_size = sizeof(metrics_vtable), \
+                        },                                              \
+                },                                                      \
+        }
+
+#define METRICS_FAMILY(_name, _description, _type, _get)                \
+        {                                                               \
+                .type = _METRICS_FAMILY,                                \
+                .x = {                                                  \
+                        .metric_family = {                              \
+                                .name = _name,                          \
+                                .description = _description,            \
+                                .type = _type,                          \
+                                .get = _get,                            \
+                        },                                              \
+                },                                                      \
+        }
+
+#define METRICS_VTABLE_END                                              \
+        {                                                               \
+                .type = _METRICS_VTABLE_END,                            \
+                .x = {                                                  \
+                        .end = {                                        \
+                                ._reserved = 0,                         \
+                        },                                              \
+                },                                                      \
+        }
+
+#define JSON_BUILD_METRIC_NAME(name) SD_JSON_BUILD_PAIR_STRING("name", name)
+#define JSON_BUILD_METRIC_FULL_NAME(name) SD_JSON_BUILD_PAIR_STRING("name", name)
+#define JSON_BUILD_METRIC_VALUE(value) SD_JSON_BUILD_PAIR_UNSIGNED("value", value)
+#define JSON_BUILD_METRIC_OBJECT(object) SD_JSON_BUILD_PAIR_STRING("object", object)
+#define JSON_BUILD_METRIC_FIELD(k, v) JSON_BUILD_PAIR_STRING_NON_EMPTY(k, v)
+#define JSON_BUILD_METRIC_FIELDS(fields) SD_JSON_BUILD_PAIR("field", SD_JSON_BUILD_OBJECT("fields", fields))
+#define JSON_BUILD_METRIC_FIELD5(k1, v1, k2, v2, k3, v3, k4, v4, k5, v5) \
+                JSON_BUILD_PAIR_STRING_NON_EMPTY(k1, v1), \
+                JSON_BUILD_PAIR_STRING_NON_EMPTY(k2, v2), \
+                JSON_BUILD_PAIR_STRING_NON_EMPTY(k3, v3), \
+                JSON_BUILD_PAIR_STRING_NON_EMPTY(k4, v4), \
+                JSON_BUILD_PAIR_STRING_NON_EMPTY(k5, v5)
```

Comment:
this should be a variadic macro `#define(, ...)`. There are plenty of examples in the code base.
for instance,
```
src/udev/udev-rules.c:409:#define log_line_debug_errno(line, error, ...)   log_line_full_errno(line, LOG_DEBUG, error, __VA_ARGS__)
```